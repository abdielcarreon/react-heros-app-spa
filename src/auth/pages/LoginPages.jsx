import { useContext } from "react";
import { useNavigate } from "react-router-dom"
import { AuthContext } from "../context";

export const LoginPages = () => {
  

  const { login } = useContext( AuthContext )
  const navigate = useNavigate();

  const onLogin = () => {

    const lastPath = localStorage.getItem('lastPath') || '/';
    
    login( 'Abdiel C' );

    navigate(lastPath, {
      replace: true
    })
  }
  return (
    <div className="container mt-5 ">
      <h1>Login</h1>
      <hr />

      <button className="btn- tbn-primary"
        onClick={ onLogin }
      >
          Login
      </button>
    </div>
  )
}

 