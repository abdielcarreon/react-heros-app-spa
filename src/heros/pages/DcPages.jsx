import { HeroList } from "../components";

 
export const DcPages = () => {
  return (
    <>
    
      <h1>DC Comics</h1>
      <hr />

      <ul>
        <HeroList publisher='DC Comics'/>
      </ul>

    </>
  )
}


