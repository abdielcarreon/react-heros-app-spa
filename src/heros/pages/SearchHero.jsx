import { useLocation, useNavigate } from "react-router-dom";

import { useForm } from "../../hook/useForm";
import { HeroCard } from "../components";
import queryString from "query-string";
import { getHerosByName } from "../helpers";

export const SearchHero = () => {

  const navigate = useNavigate();

  // Doing visible navigate(`?q=${ searchText }`). Getting the actual ubication
  const location = useLocation();

  // Installation 'query-string' to can draw search property
  const { q = '' } = queryString.parse( location.search );

  const heros = getHerosByName(q);

  const showSearch = (q.length === 0);
  const showError = (q.length > 0) && heros.length === 0;

  const { searchText, onInputChange }  = useForm({
    searchText: q
  });

  const onSearchSubmit = (e) => {
    e.preventDefault();

    // if( searchText.trim().length <= 1 ) return; 

    // Naving toward setted parameters
    navigate(`?q=${ searchText }`)
  }
  return (
    <>
      <h1>Search</h1>
      <hr />

      {/* {
        (q === '') 
        ? <div className="alert alert-primary">
          Search Hero
          </div>
          : ( heros.length === 0 ) && <div className="alert alert-danger">
          There ar en't results such <b>{ q }</b> specified
          </div>
      } */}

      

      <div className="row">
          <div className="col-5">
            <h4>Searching</h4>
            <hr />
            <form onSubmit={ onSearchSubmit }>
                <input 
                type="text" 
                placeholder="Search Hero" 
                className="form-control" 
                name="searchText"
                autoComplete="off"
                value={ searchText }
                onChange={ onInputChange }
                />

                <button className="btn btn-outline-primary mt-1">
                  Search
                </button>
            </form>
        </div>

        <div className="col-7">
            <h4>Results</h4>
            <hr />

            <div style={{ display: showSearch ? '' : 'none'}}  className="alert alert-primary animate__animated animate__fadeIn">
            Search Hero 
            </div>

            <div style={{ display: showError ? '' : 'none' }}  className="alert alert-danger animate__animated animate__fadeIn">
                There ar en't results such <b>{ q }</b> specified
            </div>


            {
              heros.map(hero => (
                <HeroCard key={ hero.id } { ...hero }/>
              ))
            }
        </div>
      </div>
    </>
  )
}

