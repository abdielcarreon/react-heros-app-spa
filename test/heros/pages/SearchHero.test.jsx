import { fireEvent, render, screen } from "@testing-library/react"
import { MemoryRouter, useNavigate } from "react-router-dom"
import { SearchHero } from "../../../src/heros/pages";


const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}));


describe('Proving <SearchPage />', () => {
  

    beforeEach(() => jest.clearAllMocks() );

    test('should display the screen by default', () => {
      
        const { container } = render(
            <MemoryRouter>
                <SearchHero />
            </MemoryRouter>
        );

        screen.debug();

        expect( container ).toMatchSnapshot();
    });

    test('should display Batman and input whit queryString value', () => {
      
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchHero />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect( input.value ).toBe('batman')

        const img = screen.getByRole('img').src
        expect(img).toContain( "/assets/heros/dc-batman.jpg" );

        //screen.debug();

        const alert = screen.getByLabelText('alert-danger')

        expect(alert.style.display).toBe('none')
    });

    test('should display an error if hero not found', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman2']}>
                <SearchHero />
            </MemoryRouter>
        );

    
        const notAlert = screen.getByLabelText( 'alert-danger' );
        expect(notAlert.className).not.toContain('display: none');

    });
    

    test('should display the "navigate" in the new screen', () => {
      
        const inputValue = 'superman';

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchHero />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change(input, { target: { name: 'searchText', value: 'superman'} })

        const form = screen.getByRole('form');
        fireEvent.submit( form )

        expect( mockedUseNavigate ).toHaveBeenCalledWith( `?q=${ inputValue }` )
    });
    
    
})
