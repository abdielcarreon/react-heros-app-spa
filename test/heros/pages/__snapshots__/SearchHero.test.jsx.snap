// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Proving <SearchPage /> should display the screen by default 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search Hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search Hero
      </div>
      <div
        aria-label="alert-danger"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        There ar en't results such 
        <b />
         specified
      </div>
    </div>
  </div>
</div>
`;
